include/Utilities.hpp

```diff
@@ -0,0 +1,398 @@
+                                  const BlockVector& fine_sol);
+
+/** @brief Print (string, double) pairs in JSON
+    @param values values to print
+    @param out output stream
+    @param pretty print each pair on its own line if true
+*/
+void PrintJSON(const std::map<std::string, double>& values, std::ostream& out = std::cout,
+               bool pretty = true);
+
+/** @brief Create aggregate to vertex relationship
+    @param partition partition of vertices
+    @returns agg_vertex aggregate to vertex relationship
+*/
+SparseMatrix MakeAggVertex(const std::vector<int>& partition);
+
+/** @brief Create processor to aggregate relationship
+    @param num_procs number of processors
+    @param num_aggs_global number of global aggregates
+    @returns proc_agg processor to aggregate relationship
+*/
+SparseMatrix MakeProcAgg(int num_procs, int num_aggs_global);
+
+/** @brief Use power iterations to find the maximum eigenpair of A
+    @param comm MPI Communicator
+    @param A operator to apply action of A
+    @param result result on output, initial guess on input
+    @param max_iter maxiumum number of iterations to perform
+    @param tol tolerance to iterate to
+    @param verbose print additional information if true
+*/
+double PowerIterate(MPI_Comm comm, const linalgcpp::Operator& A, VectorView result,
+                    int max_iter = 1000, double tol = 1e-8, bool verbose = false);
+
+/** @brief Broadcast a matrix from rank 0 to all other ranks
+    @param comm MPI Communicator
+    @param mat matrix to broadcast
+*/
+void BroadCast(MPI_Comm comm, SparseMatrix& mat);
+
+/** @brief Extract a dense submatrix from a sparse matrix
+    @param A matrix from which to extract
+    @param row row indices to extract
+    @param col column indices to extract
+    @param colMapper map of global to local indices
+    @param A_sub holds the extracted submatrix
+*/
+void ExtractSubMatrix(const SparseMatrix& A, const std::vector<int>& rows,
+                      const std::vector<int>& cols, const std::vector<int>& colMapper,
+                      DenseMatrix& A_sub);
+
+/** @brief Compute the (scaled) outer product \f$ a v v^T \f$.
+
+    @param a scalar multiple
+    @param v the vector to outer product.
+    @param aVVt the returned dense matrix.
+*/
+void MultScalarVVt(double a, const VectorView& v, DenseMatrix& aVVt);
+
+/** @brief Assemble element matrices
+
+    @param elem_dof element to dof relationship
+    @param elems set of elements to assemble
+    @returns assembled matrix
+*/
+SparseMatrix AssembleElemMat(const SparseMatrix& elem_dof, const std::vector<DenseMatrix>& elems);
+
+/** @brief Adds two sparse matrices C = alpha * A + beta * B
+
+    @param alpha scale for A
+    @param A A matrix
+    @param beta scale for B
+    @param B B matrix
+    @returns C such that C = alpha * A + beta * B
+*/
+SparseMatrix Add(double alpha, const SparseMatrix& A, double beta, const SparseMatrix& B);
```

Comment:
These three functions, `Add`, `AssembleElemMat` and `MultScalarVVt`, are also not used anywhere in the library. Some of them may be moved to the linear algebra library or just really not needed anymore? Should we remove/deprecate them?

Reply:
Yes, if they aren't being used I'll remove them.

Reply:
Looks like `Add` is used in the sparse eigensolver for `A_minus_sigma_B_inv_`, although this function should probably just be moved into `linalgcpp`.

Reply:
My bad, I was using IDE to check these functions' usage. You can decide where to put this function.